### PR TITLE
Search timeout

### PR DESCRIPTION
**Pull request by generall**
Related issue: https://github.com/qdrant/qdrant/issues/2273

- [x] Create timeout error type
- [x] Update functions signatures to handle Atomic
- [x] Store Atomic in the raw scorer
- [x] Atomic in quantized scorer
- [x] Atomic in async scorer
- [x] Add logic for terminating raw scorer operations if stopped
- [x] Add logic to switch Atomic in the async code
- [x] Configuration for timeout

---

Further work:

- [ ] Enable timeout param in API
- [ ] Refactor method calls to avoid many parameters

### CODE REVIEW

**Comment by agourlay** [comment]
That's quite the diff!

Did you test it manually on a distributed deployment as well?

**Comment by generall** [comment]
> Did you test it manually on a distributed deployment as well?

Tested on local.

File: lib/segment/src/segment.rs [hunk 1]
```diff
@@ -743,14 +744,15 @@ impl SegmentEntry for Segment {
         filter: Option<&Filter>,
         top: usize,
         params: Option<&SearchParams>,
+        is_stopped: &AtomicBool,
     ) -> OperationResult<Vec<ScoredPoint>> {
```
[the remark this answers is not in the document]
Comment (generall) [reply lib/segment/src/segment.rs:748]: done

File: lib/segment/src/segment.rs [hunk 2]
```diff
@@ -764,13 +766,14 @@ impl SegmentEntry for Segment {
         filter: Option<&Filter>,
         top: usize,
         params: Option<&SearchParams>,
+        is_stopped: &AtomicBool,
```
[the remark this answers is not in the document]
Comment (generall) [reply lib/segment/src/segment.rs:770]: done

**Review by agourlay — APPROVED** [review]

**Comment by generall** [comment]
@agourlay tested also in distributed mode and refactored a bit the error handling

**Comment by agourlay** [comment]
Refactoring LGTM!

Happy to hear you were able to test this in a distributed setup.

**Pull request merged**